### ISSUE TITLE
Documentation on Refactoring?

### ISSUE THREAD

**Issue by SystemsModelling** [rubberduck-vba/Rubberduck#2673]
I see "Extract Method", nice.
I can see you'd need to do "Move closer to usage" first.

What do all the rest do, eg the Interface items?

"Remove Parameters" could be useful, as long a we are sure they are not used.

Also: As you're parsing everything, could you create a tokenised no-comments, minimum whitespace version of the source using generic identifiers by type, and then scan it for maximum common blocks of text, to suggest refactorings?

**Comment by retailcoder** [comment]
Agreed - the wiki (and website) needs to document these features more thoroughly.

*Extract Method* currently has a number of issues though, but it's *supposed* to handle locals automatically. In other words it's *supposed* to analyze the selection and the code around it, determine what's used where (before/inside/after the selected code), determine what inputs/outputs the extracted method should have; a variable that's declared outside the selection and only used before it wouldn't be affected; a variable that's declared outside the selection and used inside but not before or after would become a local variable in the extracted method (and removed in the original); a variable that's used after the selection becomes a candidate for a return value or `ByRef` parameter, depending on how many there are.

*Remove/Reorder Parameters* changes the signature and updates all call sites. IIRC the *remove parameter* refactoring is called by the "parameter not used" inspection quickfix.

*Move closer to usage* can be used on a module-level variable that's only used in one scope (there's an inspection for that), or on a local variable - in which case the declaration is moved on the line immediately above the first reference to that variable. IIRC this needs to be fine-tuned to take `If` blocks and non-scoping loop and `With` blocks into account though.

As for the last question, that *tokenized version of the code* would be the *abstract syntax tree*, which is basically going to be our parse trees, with annotated nodes - that will unlock [[code-path-analysis]](https://github.com/rubberduck-vba/Rubberduck/labels/code-path-analysis) and open up a whole new world of possibilities, indeed!

**Comment by SystemsModelling** [comment]
[code-path-analysis]  Oooh.

**Comment by retailcoder** [comment]
Closing as completed, given refactorings now have their own individual feature records and documentation on the website.
